{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa7169f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "from gseapy import Biomart\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import os\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c79d05e8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                    Mart                    Version\n",
      "0   ENSEMBL_MART_ENSEMBL           ensembl_mart_108\n",
      "1     ENSEMBL_MART_MOUSE             mouse_mart_108\n",
      "2  ENSEMBL_MART_SEQUENCE          sequence_mart_108\n",
      "3  ENSEMBL_MART_ONTOLOGY          ontology_mart_108\n",
      "4   ENSEMBL_MART_GENOMIC  genomic_features_mart_108\n",
      "5       ENSEMBL_MART_SNP               snp_mart_108\n",
      "6   ENSEMBL_MART_FUNCGEN        regulation_mart_108\n"
     ]
    }
   ],
   "source": [
    "bm = Biomart()\n",
    "marts = bm.get_marts()\n",
    "print(marts)\n",
    "datasets = bm.get_datasets(mart='ENSEMBL_MART_ENSEMBL')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "5c7f7168",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Dataset</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>abrachyrhynchus_gene_ensembl</td>\n",
       "      <td>Pink-footed goose genes (ASM259213v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pkingsleyae_gene_ensembl</td>\n",
       "      <td>Paramormyrops kingsleyae genes (PKINGS_0.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nscutatus_gene_ensembl</td>\n",
       "      <td>Mainland tiger snake genes (TS10Xv2-PRI)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>clumpus_gene_ensembl</td>\n",
       "      <td>Lumpfish genes (fCycLum1.pri)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>llaticaudata_gene_ensembl</td>\n",
       "      <td>Blue-ringed sea krait genes (latLat_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>210</th>\n",
       "      <td>ssberkshire_gene_ensembl</td>\n",
       "      <td>Pig - Berkshire genes (Berkshire_pig_v1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>ecalabaricus_gene_ensembl</td>\n",
       "      <td>Reedfish genes (fErpCal1.1)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>pnattereri_gene_ensembl</td>\n",
       "      <td>Red-bellied piranha genes (Pygocentrus_nattere...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>neugenii_gene_ensembl</td>\n",
       "      <td>Wallaby genes (Meug_1.0)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>214</th>\n",
       "      <td>acitrinellus_gene_ensembl</td>\n",
       "      <td>Midas cichlid genes (Midas_v5)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Dataset  \\\n",
       "0    abrachyrhynchus_gene_ensembl   \n",
       "1        pkingsleyae_gene_ensembl   \n",
       "2          nscutatus_gene_ensembl   \n",
       "3            clumpus_gene_ensembl   \n",
       "4       llaticaudata_gene_ensembl   \n",
       "..                            ...   \n",
       "210      ssberkshire_gene_ensembl   \n",
       "211     ecalabaricus_gene_ensembl   \n",
       "212       pnattereri_gene_ensembl   \n",
       "213         neugenii_gene_ensembl   \n",
       "214     acitrinellus_gene_ensembl   \n",
       "\n",
       "                                           Description  \n",
       "0                Pink-footed goose genes (ASM259213v1)  \n",
       "1          Paramormyrops kingsleyae genes (PKINGS_0.1)  \n",
       "2             Mainland tiger snake genes (TS10Xv2-PRI)  \n",
       "3                        Lumpfish genes (fCycLum1.pri)  \n",
       "4             Blue-ringed sea krait genes (latLat_1.0)  \n",
       "..                                                 ...  \n",
       "210           Pig - Berkshire genes (Berkshire_pig_v1)  \n",
       "211                        Reedfish genes (fErpCal1.1)  \n",
       "212  Red-bellied piranha genes (Pygocentrus_nattere...  \n",
       "213                           Wallaby genes (Meug_1.0)  \n",
       "214                     Midas cichlid genes (Midas_v5)  \n",
       "\n",
       "[215 rows x 2 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f65b702",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Attribute</th>\n",
       "      <th>Description</th>\n",
       "      <th>Additional</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ensembl_gene_id</td>\n",
       "      <td>Gene stable ID</td>\n",
       "      <td>Stable ID of the Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ensembl_gene_id_version</td>\n",
       "      <td>Gene stable ID version</td>\n",
       "      <td>Versionned stable ID of the Gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ensembl_transcript_id</td>\n",
       "      <td>Transcript stable ID</td>\n",
       "      <td>Stable ID of the Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ensembl_transcript_id_version</td>\n",
       "      <td>Transcript stable ID version</td>\n",
       "      <td>Versionned stable ID of the Transcript</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ensembl_peptide_id</td>\n",
       "      <td>Protein stable ID</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3157</th>\n",
       "      <td>cdna_coding_start</td>\n",
       "      <td>cDNA coding start</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3158</th>\n",
       "      <td>cdna_coding_end</td>\n",
       "      <td>cDNA coding end</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3159</th>\n",
       "      <td>genomic_coding_start</td>\n",
       "      <td>Genomic coding start</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3160</th>\n",
       "      <td>genomic_coding_end</td>\n",
       "      <td>Genomic coding end</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3161</th>\n",
       "      <td>is_constitutive</td>\n",
       "      <td>Constitutive exon</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3162 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          Attribute                   Description  \\\n",
       "0                   ensembl_gene_id                Gene stable ID   \n",
       "1           ensembl_gene_id_version        Gene stable ID version   \n",
       "2             ensembl_transcript_id          Transcript stable ID   \n",
       "3     ensembl_transcript_id_version  Transcript stable ID version   \n",
       "4                ensembl_peptide_id             Protein stable ID   \n",
       "...                             ...                           ...   \n",
       "3157              cdna_coding_start             cDNA coding start   \n",
       "3158                cdna_coding_end               cDNA coding end   \n",
       "3159           genomic_coding_start          Genomic coding start   \n",
       "3160             genomic_coding_end            Genomic coding end   \n",
       "3161                is_constitutive             Constitutive exon   \n",
       "\n",
       "                                  Additional  \n",
       "0                      Stable ID of the Gene  \n",
       "1           Versionned stable ID of the Gene  \n",
       "2                Stable ID of the Transcript  \n",
       "3     Versionned stable ID of the Transcript  \n",
       "4                                             \n",
       "...                                      ...  \n",
       "3157                                          \n",
       "3158                                          \n",
       "3159                                          \n",
       "3160                                          \n",
       "3161                                          \n",
       "\n",
       "[3162 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "attrs = bm.get_attributes(dataset='hsapiens_gene_ensembl')\n",
    "attrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "eb01d867",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "164"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = pd.read_csv(\"genes.tsv\", sep = '\\t')\n",
    "len(gene_names['gene'].values.astype(str).tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "dddf6854",
   "metadata": {},
   "outputs": [],
   "source": [
    "queries ={'external_gene_name': gene_names['gene'].values.astype(str).tolist()}\n",
    "results = bm.query(dataset='hsapiens_gene_ensembl',\n",
    "                   attributes=['external_gene_name', 'entrezgene_id'],\n",
    "                   filters=queries)\n",
    "results.to_csv('enterz_id.tsv', sep = '\\t', index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e45b570d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2557e30e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
